capstone/src/main/java/com/google/sps/servlets/StudySetServlet.java

```diff
@@ -113,6 +121,102 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
       );
     }
   }
-  //TODO doPost for creating and storing a study set
 
+  public void doPost(HttpServletRequest request, HttpServletResponse response)
+    throws IOException {
+    HashMap<String, Object> result_par = new Gson()
+    .fromJson(request.getReader(), HashMap.class);
+    String owner_id = result_par.get("user_id").toString();
+    String title = result_par.get("title").toString();
+    String subject = result_par.get("subject").toString();
+    String description = result_par.get("description").toString();
+    String university = result_par.get("university_id").toString();
+    String professor = result_par.get("professor").toString();
+    String academic_time = result_par.get("academic_time").toString();
+    String course_name = result_par.get("course_name").toString();
+
+    SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+    Date date = new Date();
+    String creation_time = formatter.format(date);
+    String update_time = creation_time;
+
+    ArrayList<LinkedTreeMap<String, String>> cards = (ArrayList) result_par.get(
+      "cards"
+    );
+    String studyID = "";
+
+    ServletContext servletContext = getServletContext();
+    DataSource pool = (DataSource) servletContext.getAttribute("my-pool");
+    if (pool == null) {
+      System.err.println(
+        "Connection to SQL database not working because servlet is not conntect to the database"
+      );
+      return;
+    }
+
+    ResultSet result;
+    try (Connection conn = pool.getConnection()) {
+      String statement = String.format(
+        "INSERT INTO study_set (owner_id, title, subject, description, university_id, professor, academic_time_period, course_name, creation_time, update_time)" +
```

Comment:
Can you explain why using "?" is better? Or is it something you prefer?

Reply:
Using string concatenation (i.e String.format() ) doesn't really prevent SQL injection (even though you're still using PreparedStatements) , because it still adds the user input directly to your SQL query. But with the case of using the "?" placeholder, the user's input is treated seperately. So let's stay, eg. your SQL query is like;
String.format(SELECT * from table_name WHERE col_name = %1$s", user_input) .. if the user_input is "Caleb); DROP TABLE table_name;" , then our table will be deleted after the selecting the required output.

But if you use : sqlquery = "SELECT * FROM table_name WHERE col_name = ?", then you do sqlquery.setString(1, user_input) .... even if the user_input is  "Caleb); DROP TABLE table_name;" , it will return null (empty resultSet) since there is not a col_name with value directly equal to the string "Caleb); DROP TABLE table_name;".

Reply:
Great explanation 👍